{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Introduction\n",
    "============\n",
    "\n",
    "The field of bioinformatics exists at the intersection of biological\n",
    "research and computer science. The three key facets of the discipline\n",
    "are:\n",
    "\n",
    "1.  The creation and management of large scale datasets\n",
    "\n",
    "2.  The design of algorithms and statistical techniques aimed at teasing\n",
    "    out relationships between members of those datasets\n",
    "\n",
    "3.  The use of these tools to improve medical science, whether it be\n",
    "    drug discovery, tailoring treatments to individuals, or identifying\n",
    "    trends in public health.\n",
    "\n",
    "Much of bioinformatics revolves around genetic information. The potency\n",
    "of genetic data lies partially in its ability to uniquely identify an\n",
    "individual and provide information on various social groupings that\n",
    "individual is a member of, whether it be family, disease group, or\n",
    "ethnic group. This property allows an individual's genetic data to be\n",
    "predictive of not only their health, but the health of others. The\n",
    "analysis of genetic data thus has tremendous potency, both to help\n",
    "others and to compromise privacy. This Systematization of Knowledge\n",
    "paper aims to summarize the state of the subset of bioinformatics\n",
    "concerned with ensuring both the privacy of individuals.\n",
    "\n",
    "Secure Multiparty Computation\n",
    "=============================\n",
    "\n",
    "A de facto requirement to providing adequate privacy for bioinformatics\n",
    "applications is secure computation. In secure computation, there are two\n",
    "or more parties who each have some private data, and it is necessary to\n",
    "answer a question that relies on everyone's data. For example, in a\n",
    "vote, the majority winner must be identified. In an auction, the winning\n",
    "bid must be identified. Simultaneously, however, each party wants to\n",
    "maximize their privacy, i.e. each voter doesn't want who they voted for\n",
    "to be revealed.\n",
    "\n",
    "#### Formal Definition\n",
    "\n",
    "There are $n$ parties, where $x_i$ is the private input of the $i$th\n",
    "party, and there is a function $f(x_1, x_2, ..., x_n)$ that everyone\n",
    "wishes to compute.\n",
    "\n",
    "For the two party case, it can be shown that the output of $f$ can be\n",
    "determined without leaking any information about the private inputs for\n",
    "any arbitrary function $f$.\n",
    "\n",
    "The two parties are traditionally named Alice and Bob, where Alice will\n",
    "have secret information $a$, Bob will have secret information $b$, and\n",
    "the function $f$ is a public function whose output Alice and Bob both\n",
    "want to know without revealing additional information about $a$ or $b$\n",
    "respectively. As an example, Alice and Bob have both just received their\n",
    "MCAT scores, and want to know who scored higher on the exam without\n",
    "revealing their exact scores. Here, $a$ is Alice's score, while $b$ is\n",
    "Bob's score. The function $f$ they wish to securely compute is:\n",
    "\n",
    "$$f(a,b)= \\begin{cases}\n",
    "Alice, & \\text{if $a > b$}.\\\\\n",
    "Bob, & \\text{if $b < a$}.\\\\\n",
    "same, & \\text{otherwise}.\n",
    "\\end{cases}$$\n",
    "\n",
    "#### Security Guarantee\n",
    "\n",
    "In our imagined scenario, both Alice and Bob will adhere to the\n",
    "*Honest-but-Curious* model. In this model, neither Alice or Bob will\n",
    "lie, and both will follow the protocol. However, both Alice and Bob will\n",
    "analyze any messages sent between them and try to identify the other\n",
    "person's input.\n",
    "\n",
    "It is important to note that the assumption that Alice and Bob will not\n",
    "lie is a fairly tenuous one. It is still fairly difficult to build a\n",
    "secure protocol even in this relaxed scenario, and the\n",
    "Honest-but-Curious threat model is often used as stepping stone towards\n",
    "a stronger protocol. This paper shall ignore the existence of malicious\n",
    "adversaries as it relates to secure computation.\n",
    "\n",
    "#### Ideal World Scenario\n",
    "\n",
    "In order to evaluate the eventual solution, we will compare it to the\n",
    "best case scenario. In this ideal scenario, Alice and Bob would each\n",
    "hand their private data to a trusted third party, Faith. Faith would\n",
    "compute $f(a,b)$ and return the output to Alice and Bob without\n",
    "revealing either input.\n",
    "\n",
    "Even this ideal scenario can leak information. To see this, consider\n",
    "$f(a, b) = a + b$. Knowing the output, both Alice and Bob can determine\n",
    "the other's private data. Any proposed solution must therefore not leak\n",
    "very much more information than what can be learned with your own\n",
    "private data and the output of $f$.\n",
    "\n",
    "#### Encryption\n",
    "\n",
    "A concept that will be important to our eventual solution is that of\n",
    "*encryption*. Say you are trying to mail your friend a sensitive\n",
    "document and receive a response. Additionally, you want to ensure that\n",
    "no one can intercept either package and read the document. A solution to\n",
    "this problem would be for your friend to mail you a lock to which only\n",
    "he has the key. You place your document in a box and an additional lock\n",
    "of your own inside a box, lock it with your friend's lock, and mail it\n",
    "to your friend. Your friend receives and unlocks the box using his key.\n",
    "He then responds to your document, places it in a box, locks the box\n",
    "with your lock, and sends it. You receive the box, unlock it with your\n",
    "own key, and read the message. This protocol clearly prevents any\n",
    "eavesdroppers - the protocol of Symmetric Encryption follows the same\n",
    "structure, switching the boxes and locks for specially chosen functions.\n",
    "\n",
    "More formally, symmetric encryption involves two functions, an encryptor\n",
    "$Enc$ and a decryptor $Dec$. $Enc_k(m)$ encrypts a message $m$ with a\n",
    "key $k$, producing a ciphertext $c$. $Dec_k(c)$ decrypts the ciphertext\n",
    "$c$ using key $k$. Using the same key to decrypt as was used to decrypt\n",
    "returns back the original message: $Dec_k(Enc_k(m)) = m$. $Enc_k$ is\n",
    "chosen such that given only ciphertext $c$, it is unfeasible to find a\n",
    "key $k$ such that $Dec_k(c)$ returns the original message $m$.\n",
    "\n",
    "#### Circuits\n",
    "\n",
    "Another useful concept is that of a circuit. A circuit is simply a\n",
    "representation of a specific computation. More specifically, it is a\n",
    "non-looping sequence of operations on bits such as\n",
    "$AND, NOT, OR, XOR...$. A circuit takes its input through wires encoded\n",
    "bitwise. Likewise, it returns an output bitwise through wires. If a wire\n",
    "carries charge, it is interpreted as a $1$, $0$ if it does not. For\n",
    "every function $f(a,b)$, there exists a circuit that provides the same\n",
    "output.\n",
    "\n",
    "#### Garbled Circuits\n",
    "\n",
    "The heart of this solution to the secure computation problem is the\n",
    "garbled circuit. Informally, garbling a circuit is a method of\n",
    "\"encrypting\" the circuit representation of a function $f$ such that it\n",
    "only reveals the output of $f$ and doesn't supply any information about\n",
    "the inputs. The circuit representation of $f$ is used because the\n",
    "encryption occurs at the bitwise operation-level, garbling the\n",
    "$AND, NOT, OR, XOR..$ gates. The randomness used to encrypt the garbled\n",
    "circuit will also be used to encrypt the inputs $a$ and $b$ to ensure\n",
    "the neither Alice nor Bob learns anything about the other's input.\n",
    "\n",
    "The garbling circuit scheme proceeds as follows:\n",
    "\n",
    "1.  Alice and Bob agree on a way to represent $f$ as a circuit $C$.\n",
    "\n",
    "2.  Alice garbles the circuit $C$ yielding the garbled circuit\n",
    "    $\\hat{C}$.\n",
    "\n",
    "3.  Alice encrypts her input $a$ using the same randomness used to\n",
    "    produce $\\hat{C}$ to yield $\\hat{a}$.\n",
    "\n",
    "4.  Alice and Bob somehow arrange for Bob to encrypt his input $b$ to\n",
    "    yield $\\hat{b}$ without revealing the randomness used by Alice or\n",
    "    $b$ to Bob or Alice respectively.\n",
    "\n",
    "5.  Bob receives $\\hat{C}$ and $\\hat{a}$ and is now able to compute\n",
    "    $\\hat{C}(\\hat{a}, \\hat{b})$.\n",
    "\n",
    "6.  Bob reveals this computation to Alice.\n",
    "\n",
    "Because Alice and Bob are Honest-but-Curious, i.e. Alice doesn't\n",
    "maliciously garble some other circuit $C'$, the nature of $\\hat{C}$,\n",
    "$\\hat{a}$, and $\\hat{b}$ ensures that neither Alice or Bob learn\n",
    "anything about each other's inputs. The clear weakness in this approach,\n",
    "however, is that step 4 is not clearly described. This step can be\n",
    "achieved through oblivious transfer. The technical details of the\n",
    "oblivious transfer protocol is outside the scope of this paper, but at a\n",
    "high level, the protocol approximates the trusted third party Faith,\n",
    "ensuring that eavesdroppers learn sensitive data with low probability.\n",
    "\n",
    "Private Edit Distance\n",
    "=====================\n",
    "\n",
    "#### Introduction\n",
    "\n",
    "The Human Genome contains around 3,200,000,000 base pairs, often denoted\n",
    "as 3,200 Mb. [@naturegene] Each base pair can have one of four values:\n",
    "(A, T, C, G) and so two bits are sufficient to represent all base pairs.\n",
    "\n",
    "    A => 00\n",
    "    T => 01\n",
    "    C => 10\n",
    "    G => 11\n",
    "\n",
    "\n",
    "This means a single byte can represent a sequence of 4 base-pairs.\n",
    "Consequently, the Human Genome is about 800 megabytes long.\n",
    "Unfortunately, this is merely a theoretical lower bound. In practice,\n",
    "the average sequenced genome occupies around 200 gigabytes of disk\n",
    "space. This is a byproduct of how modern genetic sequencers work and the\n",
    "file format used to store the resulting data.[@nihfaq]\n",
    "\n",
    "This unwieldy size necessitated an optimization that takes advantage of\n",
    "the large percentage of the genome sequence that is conserved between\n",
    "individual humans. If a comparison needs to be made between two\n",
    "different genome sequences, a scientist can simply analyze how each\n",
    "genome sequence differs from a *reference genome sequence* rather than\n",
    "directly compare the base-pair sequences for each. This allows for a\n",
    "significant reduction in size. A fresh-off-the-sequencer genome is\n",
    "usually stored in a  200 GB FASTQ file, while the \"diff\" optimized\n",
    "storage method is stored in a variant (`.vcf`) file. This `.vcf` file\n",
    "usually only occupies around 125 MB which is more space efficient by\n",
    "orders of magnitude.[@vcf]\n",
    "\n",
    "An analogous idea in computer science is that of *edit distance*. Put\n",
    "simply, the edit distance between two strings is the number of\n",
    "characters that must be substituted, inserted, or deleted to change one\n",
    "of the strings into the other. It is often useful to calculate the edit\n",
    "distance between two different genomes. If a patient is diagnosed with a\n",
    "disease, a physician may want to see the prognosis of other genetically\n",
    "similar patients with the disease.\n",
    "\n",
    "From both a practical and privacy standpoint, it is undesirable to send\n",
    "a patient's entire genome sequence whenever a comparison is required.\n",
    "Because each genome sequence is 200 gigabytes in size, bandwidth will\n",
    "quickly become a bottleneck. Additionally, genome sequences are\n",
    "practically unparalleled in their ability to uniquely identify an\n",
    "individual. Furthermore, they contain sensitive medical data - not only\n",
    "the prescence of congenital diseases but predispositions to a wide swath\n",
    "of additional maladies.\n",
    "\n",
    "The .vcf \"diff\" method only alleviates one of these problems - the one\n",
    "concerning bandwidth. Because the human reference genome sequence is\n",
    "publicly available, reverse engineering an individual's genome sequence\n",
    "given a comprehensive variant file is trivial - simply start with the\n",
    "reference and make the changes described in the variant file. Similarly,\n",
    "if you are given a variant file that describes the set difference\n",
    "between your genome sequence and another person $Y$'s sequence, the\n",
    "derivation of $Y$'s sequence is trivial as well.\n",
    "\n",
    "#### Problem Statement\n",
    "\n",
    "Clearly, we cannot allow this set difference to be calculated. However,\n",
    "if we take advantage of three features we can reduce the problem of\n",
    "calculating the set difference while maintaining privacy to the problem\n",
    "of calculating the *size* of the set difference while maintaining\n",
    "privacy.\n",
    "\n",
    "1.  Most differences between human genome sequences are *substitutions*,\n",
    "    not insertions or deletions.\n",
    "\n",
    "2.  There exists a public reference genome and many variations that have\n",
    "    already been computed.\n",
    "\n",
    "3.  Using probabilistic algorithms, the size of the set difference can\n",
    "    be securely approximated while keeping the actual set difference\n",
    "    private.\n",
    "\n",
    "The algorithm is as follows:\n",
    "\n",
    "1.  Party A calculates the minimum edit sequences from the reference\n",
    "    genome to genome A using the levenshtein distance.\n",
    "\n",
    "2.  Party B calculates the minimum edit sequences from the reference\n",
    "    genome to genome B using the levenshtein distance.\n",
    "\n",
    "3.  The parties run a secure computation protocol to approximate the\n",
    "    cardinality of the set difference of the minimum edit sequences.\n",
    "\n",
    "This third step has its own multi-step protocol. While the technical\n",
    "details can be found in the implementation included with this report, at\n",
    "a high level the protocol works by \"squeezing\" each set of edit\n",
    "sequences into an integer. This \"squeezing\" is performed by taking the\n",
    "sum of the binary hash ($h : edit \\rightarrow \\{-1, 1\\}$) of each edit\n",
    "in the set. Once the \"squeezed\" values $d_A$ and $d_B$ are computed, the\n",
    "square of the difference is securely computed $(d_A - d_B)^2$. This\n",
    "squeezing process is performed $l$ times, and then the $k$-median is\n",
    "securely computed. Here, $l$ and $k$ are used to bound the accuracy and\n",
    "the efficiency.\n",
    "\n",
    "The above algorithm is implemented in `sec_ped.py` contained in the\n",
    "`demos` directory (where all other demos will be located). As a\n",
    "comparison for the accuracy of the secure computation demo, an unsecure\n",
    "algorithm is implemented in `unsec_ped.py`.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_min_edits(ref_str, trans_str):\n",
    "    leven = levenshtein(ref_str, trans_str)\n",
    "    edits = leven[1]\n",
    "    parsed = parse_edits(edits, ref_str, trans_str)\n",
    "    return parsed\n",
    "    \n",
    "def parse_edits(edits, ref_str, trans_str):\n",
    "    parsed = []\n",
    "    for line in edits:\n",
    "        op = line['type']\n",
    "        loc = int(line['i']) + 1\n",
    "        trans_loc = int(line['j']) - 1\n",
    "        change = trans_str[trans_loc]\n",
    "        if(op == \"deletion\"):\n",
    "            parsed.append(str(loc) + op[0:3] + '')\n",
    "        elif(op == \"insertion\" or  op == \"substitution\"):\n",
    "            parsed.append(str(loc) + op[0:3] + change)\n",
    "    return parsed\n",
    "\n",
    "def compute_edit_set_diff(A_edits, B_edits):\n",
    "    set_diff = []\n",
    "    for x in A_edits:\n",
    "        if x not in B_edits:\n",
    "            set_diff.append(x)\n",
    "    for x in B_edits:\n",
    "        if x not in A_edits:\n",
    "            set_diff.append(x)\n",
    "\n",
    "    return set_diff\n",
    "\n",
    "# Calculates the levenshtein distance and the edits between two strings\n",
    "def levenshtein(ref_str, op_str, key=hash):\n",
    "  rows = costmatrix(ref_str, op_str, key)\n",
    "  edits = backtrace(ref_str, op_str, rows, key)\n",
    " \n",
    "  return rows[-1][-1], edits\n",
    " \n",
    "# Generate the cost matrix for the two strings\n",
    "def costmatrix(ref_str, op_str, key=hash):\n",
    "  rows = []\n",
    " \n",
    "  previous_row = range(len(op_str) + 1)\n",
    "  rows.append(list(previous_row))\n",
    " \n",
    "  for i, c1 in enumerate(ref_str):\n",
    "    current_row = [i + 1]\n",
    "    for j, c2 in enumerate(op_str):\n",
    "      insertions = previous_row[j + 1] + 1\n",
    "      deletions = current_row[j] + 1\n",
    "      substitutions = previous_row[j] + (key(c1) != key(c2))\n",
    "      current_row.append(min(insertions, deletions, substitutions))\n",
    "    previous_row = current_row\n",
    " \n",
    "    rows.append(previous_row)\n",
    " \n",
    "  return rows\n",
    " \n",
    "# Trace back through the cost matrix to generate the list of edits\n",
    "def backtrace(ref_str, op_str, rows, key=hash):\n",
    "  i, j = len(ref_str), len(op_str)\n",
    " \n",
    "  edits = []\n",
    " \n",
    "  while(not (i == 0  and j == 0)):\n",
    "    prev_cost = rows[i][j]\n",
    " \n",
    "    neighbors = []\n",
    " \n",
    "    if(i!=0 and j!=0):\n",
    "      neighbors.append(rows[i-1][j-1])\n",
    "    if(i!=0):\n",
    "      neighbors.append(rows[i-1][j])\n",
    "    if(j!=0):\n",
    "      neighbors.append(rows[i][j-1])\n",
    " \n",
    "    min_cost = min(neighbors)\n",
    " \n",
    "    if(min_cost == prev_cost):\n",
    "      i, j = i-1, j-1\n",
    "      edits.append({'type':'match', 'i':i, 'j':j})\n",
    "    elif(i!=0 and j!=0 and min_cost == rows[i-1][j-1]):\n",
    "      i, j = i-1, j-1\n",
    "      edits.append({'type':'substitution', 'i':i, 'j':j})\n",
    "    elif(i!=0 and min_cost == rows[i-1][j]):\n",
    "      i, j = i-1, j\n",
    "      edits.append({'type':'deletion', 'i':i, 'j':j})\n",
    "    elif(j!=0 and min_cost == rows[i][j-1]):\n",
    "      i, j = i, j-1\n",
    "      edits.append({'type':'insertion', 'i':i, 'j':j})\n",
    " \n",
    "  edits.reverse()\n",
    " \n",
    "  return edits\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading ref genome from: ./genome_ref.txt\n",
      "Reading genome A from: ./genome_A.txt\n",
      "Reading genome B from: ./genome_B.txt\n",
      "|A - B|: 936\n"
     ]
    }
   ],
   "source": [
    "def main():\n",
    "    print(\"Reading ref genome from: ./genome_ref.txt\")\n",
    "    ref_gen = open(\"./genome_ref.txt\", \"r\").read()\n",
    "    print(\"Reading genome A from: ./genome_A.txt\")\n",
    "    A_gen = open(\"./genome_A.txt\", \"r\").read()\n",
    "    print(\"Reading genome B from: ./genome_B.txt\")\n",
    "    B_gen = open(\"./genome_B.txt\", \"r\").read()\n",
    "    B_min_edits = compute_min_edits(B_gen, ref_gen)\n",
    "    A_min_edits = compute_min_edits(A_gen, ref_gen)\n",
    "    A_B_diff = compute_edit_set_diff(A_min_edits, B_min_edits)\n",
    "    print(\"|A - B|: \" + str(len(A_B_diff)))\n",
    "    \n",
    "#don't worry about forward declaring\n",
    "if __name__==\"__main__\":\n",
    "   main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading ref genome from: ./genome_ref.txt\n",
      "Reading genome A from: ./genome_A.txt\n",
      "Reading genome B from: ./genome_B.txt\n",
      "Enter privacy budget epsilon: 0.1\n",
      "|A - B|: 1101.6464646464647\n"
     ]
    }
   ],
   "source": [
    "import mmh3\n",
    "import statistics\n",
    "import random\n",
    "from progress.bar import Bar\n",
    "\n",
    "'''\n",
    "Input of A: a set A.\n",
    "Input of B: a set B.\n",
    "Public Input: a (sufficiently long) common random string\n",
    "\n",
    "k and l are accuracy bounds\n",
    "\n",
    "Output: |Diff(A, B)| is calculated as the k-median of l-mean of random sampling of (d_A - d_B)^2\n",
    "\n",
    "1. For j in (1..k)\n",
    "   a. For i in (1..l)\n",
    "      i. A and B use the (same) random string to randomly pick a function h from the family of hash functions.\n",
    "      ii. A computes d_A = Sum_{s \\in A} h(s) and B computes d_b = Sum_{s \\in B} h(s), both independently.\n",
    "      iii. A and B securely compute D_i = (d_A - d_b)^2\n",
    "   b. A and B securely compute D^_j = Sum(i -> l) D_i / l\n",
    "2. A and B securely compute the median Z of D^_1 ... D^_k\n",
    "3. Output Z\n",
    "'''\n",
    "def main():\n",
    "    print(\"Reading ref genome from: ./genome_ref.txt\")\n",
    "    ref_gen = open(\"./genome_ref.txt\", \"r\").read()\n",
    "    print(\"Reading genome A from: ./genome_A.txt\")\n",
    "    A_gen = open(\"./genome_A.txt\", \"r\").read()\n",
    "    print(\"Reading genome B from: ./genome_B.txt\")\n",
    "    B_gen = open(\"./genome_B.txt\", \"r\").read()\n",
    "    A_min_edits = compute_min_edits(A_gen, ref_gen)\n",
    "    B_min_edits = compute_min_edits(B_gen, ref_gen)\n",
    "    \n",
    "    pub_rand_str = [\"19528620395885535\", \"ajajcwdeiqoksmzlhtes\", \"jha\", \"mahmoody\", \"cryptography\"]\n",
    "    epsilon = float(input(\"Enter privacy budget epsilon: \") or \"0.1\")\n",
    "    l = int(1/(epsilon**2))\n",
    "    k = 20\n",
    "    D_hat = []\n",
    "    bar = Bar('Computing...     ', max=k) \n",
    "    for j in range(1,k+1):\n",
    "        D = []\n",
    "        for i in range(1,l+1):\n",
    "            seed = get_seed(pub_rand_str[random.randint(0,4)])\n",
    "            d_A = compute_d_X(A_min_edits, seed)\n",
    "            d_B = compute_d_X(B_min_edits, seed)\n",
    "            D_i = sec_compute_D_i(d_A, d_B)\n",
    "            D.append(D_i)\n",
    "        D_hat_j = sec_compute_D_hat_j(D, l)\n",
    "        D_hat.append(D_hat_j)\n",
    "        bar.next()\n",
    "    median = statistics.median(D_hat)\n",
    "    bar.finish()\n",
    "    print(\"|A - B|: \" + str(median))\n",
    "\n",
    "def get_seed(str):\n",
    "    return mmh3.hash(str, random.randint(1, 1000))\n",
    "\n",
    "def compute_d_X(setX, seed):\n",
    "    sum = 0\n",
    "    for s in setX:\n",
    "        hashed = mmh3.hash(s, seed)\n",
    "        if(hashed % 2 == 0):\n",
    "            sum += 1\n",
    "        else:\n",
    "            sum += -1\n",
    "    return sum\n",
    "\n",
    "##FUTURE WORK: implement secure compute\n",
    "def sec_compute_D_i(d_A, d_B):\n",
    "    val = (d_A - d_B)**2\n",
    "    return val\n",
    "\n",
    "##FUTURE WORK: implement secure compute\n",
    "def sec_compute_D_hat_j(D, l):\n",
    "    sum = 0\n",
    "    for i in range(0,l):\n",
    "        val = D[i] / l\n",
    "        sum += val\n",
    "    return sum\n",
    "\n",
    "#don't worry about forward declaring\n",
    "if __name__==\"__main__\":\n",
    "   main()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This approximation process's accuracy is aided by the special\n",
    "distribution of human genome sequences.[@naturegene] For any two\n",
    "suitably unrelated individuals:\n",
    "\n",
    "1.  An overwhelming majority of their sequences are conserved, at least\n",
    "    99%\n",
    "\n",
    "2.  The places at which they differ, i.e. the location of their edits\n",
    "    from a reference genome are not close together.\n",
    "\n",
    "3.  A majority of the differences between genome sequences are\n",
    "    substitutions, around 80%\n",
    "\n",
    "#### Alternatives\n",
    "\n",
    "An interesting alternate approach was proposed by Cho et al. in their\n",
    "paper *Secure genome-wide association analysis using multiparty\n",
    "computation*. In it, they discuse the \"Beaver Multiplication Triples\"\n",
    "optimization which allows for efficient secure matrix multiplication -\n",
    "an operation that features prominently in genome comparisons. An\n",
    "additional optimization reduced the communication cost through the use\n",
    "of pseudorandom generators. As opposed to the \"squeezing\" process\n",
    "described above, a pseudorandom generator is an algorithm that generates\n",
    "numbers whose properties are similar to a sequence of random numbers.\n",
    "This optimization reduces communication i/o as both parties can obtain\n",
    "numbers from the PRG without interacting.[@natureopt]\n",
    "\n",
    "Differential Privacy\n",
    "====================\n",
    "\n",
    "#### Introduction\n",
    "\n",
    "In the previous section, the utility and risks of sharing an\n",
    "individual's genome, was discussed. This discussion applies to personal\n",
    "medical information in general, whether it be the gut microbiome, the\n",
    "proteome, or simply a medical history. When collecting, storing, and\n",
    "querying this personal data, the privacy of the individuals must be\n",
    "maintained, especially as the ubiquity of collection and analysis\n",
    "increases. More specifically, it is a vital requirement that systems\n",
    "that store sensitive data do not allow adversaries to learn about the\n",
    "presence or even absence of an individual's data in a given dataset.\n",
    "\n",
    "#### Some Straw-man Solutions\n",
    "\n",
    "To convey the non-triviality of this problem, two common-sense\n",
    "approaches will be briefly introduced and examined.\n",
    "\n",
    "1.  **Only Group Queries** If individual privacy is the concern, why not\n",
    "    only allow group-level queries? As an example, consider the\n",
    "    group-level query: \\\"How many patients in this ward have diabetes?\\\"\n",
    "    An easy way to violate the privacy of patient A is to follow up the\n",
    "    previous query with: \\\"How many patients aside from patient A have\n",
    "    diabetes?\\\" If the number from the first query is greater than the\n",
    "    number from the second, patient A must have diabetes, and must not\n",
    "    otherwise. Clearly, the privacy of patient A has been compromised.\n",
    "\n",
    "2.  **Random Noise** Why not obfuscate the data returned by introducing\n",
    "    random noise to the output? With repeated queries, the true value of\n",
    "    the output could be determined statistically.\n",
    "\n",
    "#### Problem Statement\n",
    "\n",
    "In order to propose a solution that is superior to those described\n",
    "above, the problem must be formally stated. At a high level,\n",
    "Differential Privacy is the requirement that the result of a query\n",
    "should not reveal the prescence or abscence of any individual record in\n",
    "the input dataset. Informally, the outcome of a query should be nearly\n",
    "equally likely if the dataset has your information in it, as to if it\n",
    "did not.\n",
    "\n",
    "#### Formal Definition\n",
    "\n",
    "Formally, given a dataset $A$ and $B$, $A$ and $B$ are said to be\n",
    "*adjacent* if there is a single record $r$ that is in one but not the\n",
    "other. For adjacent datasets $A$ and $B$, a query $Q$ has\n",
    "$\\epsilon$-differential privacy if for any adjacent datasets $A$ and\n",
    "$B$, and any subset $C$ of possible outcomes $Range(Q)$,\n",
    "$\\Pr[Q(A) \\in C] \\leq \\exp(\\epsilon) \\times \\Pr[M(B) \\in C]$\n",
    "\n",
    "#### Laplace Noise\n",
    "\n",
    "A function $f$ can be made to satisfy $\\epsilon$-differential privacy by\n",
    "adding Laplace noise to it. This noise takes the form of a variable $L$\n",
    "that follows the Laplace Distribution.\n",
    "\n",
    "\\bigbreak\n",
    "$f(x) + L$ where $L = Laplace(0, \\sigma)$,\n",
    "$\\sigma \\geq \\frac{\\Delta f}{\\epsilon}$, and\n",
    "$\\Delta f = \\max(||f(x) - f(x')||_1)$\n",
    "\n",
    "\\bigbreak\n",
    "The PDF for the Laplace Distribution is defined as:\n",
    "$\\frac{1}{2b} \\times \\exp(-\\frac{|x - \\mu|}{b})$ where $b$ is a scaling\n",
    "factor. The Manhattan norm is defined as:\n",
    "$||x||_1 = \\frac{1}{n} \\sum_1^n |x_1|$.\n",
    "\n",
    "As an example, consider the function $f$ that computes the average\n",
    "weight of a patient cohort. Now consider two adjacent datasets of size\n",
    "$4$ $P1$ and $P2$ that differ by only a single patient's data.\n",
    "$P1 = \\{110, 130, 145, 210\\}$ and $P2 = \\{110, 130, 145, 189\\}$ We will\n",
    "calculate the average weight of $P1$ while ensuring\n",
    "$\\epsilon$-differential privacy. As the minimum weight is $110$ and the\n",
    "maximum is $210$, $\\Delta f = \\frac{210 - 110}{4} = 25$. Adding the term\n",
    "$L$, where $L$ is the result of sampling\n",
    "$Laplace(0, \\frac{25}{\\epsilon})$, would preserve the privacy of the\n",
    "patient that is in $P1$ but not $P2$.\n",
    "\n",
    "#### Extending Differential Privacy\n",
    "\n",
    "Any algorithm that satisfies $\\epsilon$-differential privacy can be\n",
    "augmented to ensure privacy for a group of size $s$ simply by scaling\n",
    "the privacy budget $\\epsilon$ by $s$.\n",
    "\n",
    "#### Limitations\n",
    "\n",
    "It is important to note the limitations of the above algorithm,\n",
    "particularly when it is applied to personalized medicine. Personalized\n",
    "medicine can be used to guide doctors in calculating dosages that are\n",
    "tailored to individuals. A strong candidate for personalized dosing is\n",
    "Warfarin, a popular anticoagulant. Both underdoses and overdoses can\n",
    "lead to death, underscoring the importance of finding the correct dose.\n",
    "\n",
    "![Mortality and $\\epsilon$](./figures/warfarin.png)\n",
    "\n",
    "Using a dataset provided by the *International Warfarin Pharmacogenetics\n",
    "Consortium* (IWPC), researchers at the University of Wisconsin\n",
    "constructed linear regression models designed to guide doctors in\n",
    "dosing. They then proceded to perform model inversion attacks to\n",
    "determine how much sensitive data these models would leak. These *model\n",
    "inversion attack* would take the model and try to predict the presence\n",
    "or absence of specific genes in the individual genome.[@warfarin]\n",
    "Differential privacy would effectively combat these types of attacks,\n",
    "but at the risk of clinical efficacy. In fact, differential privacy\n",
    "became effective at $\\epsilon \\leq 1$, but the risk of either mortality\n",
    "or morbidity increased significantly, as can be seen in the figure\n",
    "above. Clearly the effectiveness of Differential Privacy is bounded in\n",
    "by the level of acceptable increase in mortality.\n",
    "\n",
    "Conclusion\n",
    "==========\n",
    "\n",
    "Genetic data has the capacity to both uniquely identify and reveal\n",
    "sensitive information about specific individuals and those related to\n",
    "them. The increase in the capabilities of the field bioinformatics has\n",
    "been met with increasing concern over genetic privacy. The malicious use\n",
    "of genetic data can lead to discrimination, exploitation, and identity\n",
    "theft. Gene data can also be stored indefinitely, and future techniques\n",
    "may tease out even stronger analytics than can be produced today. For\n",
    "these reasons, privacy in bioinformatics is a tremendously important\n",
    "topic in the field now and for the forseeable future.\n",
    "\n",
    "\\bibitem{naturegene}\n",
    "[Initial sequencing and analysis of the human\n",
    "genome](https://www.nature.com/articles/35057062), Nature\n",
    "\n",
    "\\bibitem{nihfaq}\n",
    "[Genomic\n",
    "Science](https://www.genome.gov/19016904/faq-about-genetic-and-genomic-science/),\n",
    "National Institues of Health\n",
    "\n",
    "\\bibitem{vcf}\n",
    "[Variant Call\n",
    "Files](https://www.internationalgenome.org/wiki/Analysis/vcf4.0),\n",
    "International Genome\n",
    "\n",
    "\\bibitem{editdistance}\n",
    "[A Guided Tour to Approximate String\n",
    "Matching](https://repositorio.uchile.cl/bitstream\\-/handle/2250/126168/Navarro_Gonzalo\\-_Guided_tour.pdf),\n",
    "University of Chile\n",
    "\n",
    "\\bibitem{warfarin}\n",
    "[Privacy in\n",
    "Pharmacogenetics](https://www.usenix.org/system/files/conference/usenixsecurity14/sec14-paper-fredrikson-privacy.pdf),\n",
    "USENIX 2014\n",
    "\n",
    "\\bibitem{natureopt}\n",
    "[Secure genome-wide association analysis using multiparty\n",
    "computation](https://www.nature.com/articles/nbt.4108), Nature\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
